### PR TITLE
Id embedding training loop heterogeneous

### PR DESCRIPTION
**Pull request by swong3-sc**
**Scope of work done**

This is my training loop for LightGCN on the gowalla dataset.
There are a number of issues I can't seem to figure out:

1. The dataloaders seem to fail when training (usually at the 17th batch). There is no error it just hangs at line `587`
2. The loss doesn't decrease

I have tried many things to address the loss:

1. Adding a sparse optimizer from TorchRec (RowWiseAdagrad)
2. Added Xavier uniform initialization after discovering the initial embeddings were too small, and thus would not propagate.
3. After the above didn't work, decided to scale up the embeddings after the forward pass, under the assumption that the convolution was shrinking the embeddings too much

I use this command to run the training locally:
`PYTHONWARNINGS=ignore WORLD_SIZE=1 RANK=0 MASTER_ADDR="localhost" MASTER_PORT=20000 python -m GiGL.examples.id_embeddings.heterogeneous_training --task_config_uri="gs://gigl-perm-dev-assets/swong3_gowalla_data_preprocessor_24/config_populator/frozen_gbml_config.yaml"`

The parameters are derived from the task config, but I will list them here for reference:
`Got training args local_world_size=1,         num_neighbors={user-to_train-item: [10, 10], user-to_test-item: [0, 0], item-to_train-user: [15, 15], item-to_test-user: [0, 0]},         sampling_workers_per_process=4,         main_batch_size=16,         random_batch_size=16,         embedding_dim=64,         num_layers=2,         num_random_negs_per_pos=1,         l2_lambda=0.0,         sampling_worker_shared_channel_size=4GB,         process_start_gap_seconds=0,         log_every_n_batch=50,         learning_rate=0.01,         weight_decay=0.0005,         num_max_train_batches=1000,         num_val_batches=100,         val_every_n_batch=50`

Where is the documentation for this feature?: N/A

Did you add automated tests or write a test plan?

***Updated Changelog.md?*** NO

***Ready for code review?:*** NO

### CODE REVIEW

File: examples/id_embeddings/heterogeneous_training.py [hunk 1]
```diff
@@ -0,0 +1,936 @@
...
+
+    loss = bpr_loss(q, pos, neg, l2_lambda=l2_lambda, l2_params=None)
+
+    if debug_log:
+        debug_info["bpr_loss"] = loss.item()
+
+    return loss, debug_info
+
+
+def _training_process(
```
Comment (nshah-sc) [inline examples/id_embeddings/heterogeneous_training.py]: i would recommend testing everything related to loss reduction in single-process, no distributed

**Comment by nshah-sc** [comment]
can you add a run command so we know all the parameter used and how to kick this off locally?

File: examples/id_embeddings/heterogeneous_training.py [hunk 2]
```diff
@@ -0,0 +1,936 @@
...
+    torch.distributed.barrier()
+
+    return main_loader, random_negative_loader
+
+
+def bpr_loss(
+    query_emb: torch.Tensor,  # [M, D]
+    pos_emb: torch.Tensor,  # [M, D]
+    neg_emb: torch.Tensor,  # [M, D] or [M, K, D]
+    l2_lambda: float = 0.0,
```
[the remark this answers is not in the document]
Comment (swong3-sc) [reply examples/id_embeddings/heterogeneous_training.py]: I will look into disabling this and comparing the output.

File: examples/id_embeddings/heterogeneous_training.py [hunk 3]
```diff
@@ -0,0 +1,936 @@
...
+    # This is the correct TorchRec pattern - optimizer must be applied before sharding
+    logger.info(f"---Rank {rank} applying sparse optimizer to embedding bag collection (BEFORE DMP)")
+    sparse_lr = learning_rate
+    for name, param in base_model._embedding_bag_collection.named_parameters():
+        logger.info(f"  Applying RowWiseAdagrad to {name}")
+        apply_optimizer_in_backward(
+            optimizer_class=RowWiseAdagrad,
+            params=[param],
+            optimizer_kwargs={"lr": sparse_lr, "weight_decay": weight_decay},
+        )
```
[the remark this answers is not in the document]
Comment (swong3-sc) [reply examples/id_embeddings/heterogeneous_training.py]: Will change

File: examples/id_embeddings/heterogeneous_training.py [hunk 4]
```diff
@@ -0,0 +1,936 @@
...
+        # Xavier uniform: U(-sqrt(6/(fan_in + fan_out)), sqrt(6/(fan_in + fan_out)))
+        # For embeddings: fan_in = 1, fan_out = embedding_dim
+        # Then scale up to combat LightGCN's aggressive neighbor averaging
+        torch.nn.init.xavier_uniform_(param)
+        param.data *= EMBEDDING_SCALE
+        init_count += 1
+
+        logger.info(f"    AFTER init (scaled {EMBEDDING_SCALE}x) - mean={param.mean().item():.6f}, std={param.std().item():.6f}, norm={param.norm().item():.6f}")
+
+    logger.info(f"Initialized {init_count} embedding parameters after DMP wrapping with {EMBEDDING_SCALE}x scaling")
```
[the remark this answers is not in the document]
Comment (swong3-sc) [reply examples/id_embeddings/heterogeneous_training.py]: Yeah, I agree
